packages/theme-language-server-common/src/documentLinks/DocumentLinksProvider.ts

```diff
@@ -91,3 +129,149 @@ function range(textDocument: TextDocument, node: { position: LiquidHtmlNode['pos
 function isLiquidString(node: LiquidHtmlNode): node is LiquidString {
   return node.type === NodeTypes.String;
 }
+
+function createDocumentLinkForTypeNode(
+  typeNode: JSONNode,
+  parentNode: LiquidRawTag,
+  textDocument: TextDocument,
+  root: URI,
+  blockType: string,
+): DocumentLink | null {
+  const startOffset = typeNode.offset;
+  const endOffset = typeNode.offset + typeNode.length;
+  const startPos = parentNode.body.position.start + startOffset;
+  const endPos = parentNode.body.position.start + endOffset;
+
+  const start = textDocument.positionAt(startPos);
+  const end = textDocument.positionAt(endPos);
+
+  return DocumentLink.create(
+    Range.create(start, end),
+    Utils.resolvePath(root, 'blocks', `${blockType}.liquid`).toString(),
+  );
+}
+
+function processPresetBlocks(
+  blocksNode: JSONNode,
+  parentNode: LiquidRawTag,
+  textDocument: TextDocument,
+  root: URI,
+): DocumentLink[] {
+  const links: DocumentLink[] = [];
+
+  if (blocksNode.type === 'object' && blocksNode.children) {
+    blocksNode.children.forEach((propertyNode) => {
+      const blockValueNode = propertyNode.children?.[1]; // The value node of the property
+      if (!blockValueNode) return;
+
+      // Check if the block has a 'name' key so we don't deeplink inline block types
+      const nameNode = findNodeAtLocation(blockValueNode, ['name']);
+      if (nameNode) {
+        return;
+      }
+
+      const typeNode = findNodeAtLocation(blockValueNode, ['type']);
+      if (typeNode && typeNode.type === 'string' && typeof typeNode.value === 'string') {
```

Comment:
When evaluating a generic variable conforms to a specific structure, you can do something called "type narrowing" when there is a specific type definition for the structure of data that you're verifying for. This has typescript benefits of showing proper typing feedback as you write code inside these conditionals. Take a look at these examples of type narrowing code in the theme-tools repo: https://github.com/Shopify/theme-tools/blob/deeplink-theme-blocks/packages/theme-check-common/src/types.ts#L23-L25